addons/hr_recruitment/models/hr_job.py

```diff
@@ -13,14 +13,23 @@ class Job(models.Model):
 
     @api.model
     def _default_address_id(self):
-        return self.env.company.partner_id
+        last_used_address = self.env['hr.job'].search([('company_id', 'in', self.env.companies.ids)], order='id desc', limit=1)
+        if last_used_address:
+            return last_used_address.address_id
+        else:
+            return self.env.company.partner_id
+
+    def _address_id_domain(self):
+        return ['|', '&', '&', ('type', '!=', 'contact'), ('type', '!=', 'private'),
+                ('id', 'in', self.sudo().env.companies.partner_id.child_ids.ids),
```

Comment:
This would allow to select an address from a company different than the job's

Maybe instead you can add a related field to company_id.partner_id then use it in a domain like:
`domain="[('type', '!=', 'contact'), ('type', '!=', 'private'), ('id', 'child of', company_partner_id)]"`